dask/array/tests/test_routines.py

```diff
@@ -39,27 +49,55 @@ def test_array():
     ((4, 6, 8), (2, 3, 4)),
     ((4, 6, 8, 10), (2, 3, 4, 5)),
 ])
-def test_atleast_nd(funcname, num_arrs, shape, chunks):
+def test_atleast_nd_one_arg(funcname, shape, chunks):
     np_a = np.random.random(shape)
     da_a = da.from_array(np_a, chunks=chunks)
 
-    np_a_n = num_arrs * [np_a]
-    da_a_n = num_arrs * [da_a]
+    np_func = getattr(np, funcname)
+    da_func = getattr(da, funcname)
+
+    np_r = np_func(np_a)
+    da_r = da_func(da_a)
+
+    assert_eq(np_r, da_r)
+
+
+@pytest.mark.parametrize("funcname", [
+    "atleast_1d",
+    "atleast_2d",
+    "atleast_3d",
+])
+@pytest.mark.parametrize("shape1, chunks1", [
+    (tuple(), tuple()),
+    ((4,), (2,)),
+    ((4, 6), (2, 3)),
+    ((4, 6, 8), (2, 3, 4)),
+    ((4, 6, 8, 10), (2, 3, 4, 5)),
+])
+@pytest.mark.parametrize("shape2, chunks2", [
+    (tuple(), tuple()),
+    ((4,), (2,)),
+    ((4, 6), (2, 3)),
+    ((4, 6, 8), (2, 3, 4)),
+    ((4, 6, 8, 10), (2, 3, 4, 5)),
+])
+def test_atleast_nd_two_args(funcname, shape1, chunks1, shape2, chunks2):
```

Comment:
My concern was the large number of tests for relatively simple functions. This has replaced what was 45 tests (which I thought were too many) with 93 tests.

Can you try out just a few combinations here for both shape1 and shape2? Perhaps:

```python
# Some smaller set of combinations
@pytest.mark.parametrize("shape1, shape2")
                         [((), (4,)),
                          ((4,6), (4, 6, 8))
                          ((4, 6, 8, 10,), ())])
def test_atleast_nd_two_args(funcname, shape1, shape2):
    np_a_1 = np.random.random(shape1)
    # don't need to parametrize chunks, using an integer works fine
    # since the chunking doesn't matter here
    da_a_1 = da.from_array(np_a_1, chunks=2)

    np_a_2 = np.random.random(shape2)
    da_a_2 = da.from_array(np_a_2, chunks=2)

    ...
```

My point is that you really don't need to test the cartesian product of all of these combinations to have a sufficient test.

Reply:
It's true there are more tests on the two argument front. However there are less in the zero argument case, which seemed to be the contention. Have added more tests for the two argument case as the previous test was not doing a good job of covering heterogeneous array shapes.

While I agree these are relatively simple functions, they are often the backbone of any codebase like Dask. As proof of this, `atleast_2d` and `atleast_3d` have been around since 0.7.6 (added in PR ( https://github.com/dask/dask/pull/854 )). If they perform poorly, weird behavior will bubble up through the codebase. The easiest way to be sure these functions perform well is to try a wide range of arguments (maybe even with some duplicates) to be really sure we are not missing some special case.

If the real concern is testing time, these tests actually run quite fast. Compared to 0.15.4, running `test_routines.py` with these changes takes ~0.3s longer. This feels pretty negligible to me. Though maybe there are concerns outside of testing time.

That said, we could constrain the second array to be bigger or equal in size to first array. That should roughly halve the number of tests for the two argument case. Thoughts?

Reply:
The real concern is testing time. Even if these specific tests run fast, if extra unnecessary tests are added in every PR eventually they will sum to a non-negligible measure.

All I'm asking for here is to consider the function(s) being tested and curate the set of test cases into a subset that specifically targets the behavior(s) you're trying to ensure. Testing with a few heterogeneous shapes may be necessary to ensure correct behavior, but you don't need 25 cases.

I'll cede to your judgement here. Feel free to pare this down or merge as is.

Reply:
I understand. Should add that I do appreciate the nudge to do so. With the slimming down of the no argument tests and now the slimming down of the two argument cases (included in the latest commit), I think we have done this.

Reply:
I don't think I'd use `pytest.skip` to do this. To me skip indicates a test that *could* run given a different setup. Instead I'd filter the cases to parametrize beforehand. Perhaps:

```python
from itertools import product

 @pytest.mark.parametrize("shape1, shape2",
     [(s1, s2) for s1, s2 in
      product(*[((), (4,), (4, 6), (4, 6, 8), (4, 6, 8, 10))]*2)
      if len(s1) >= len(s2)])
```
and then fix `chunks=2` in `from_array` (as said above, the chunking doesn't really matter here).

Up to you though.

Reply:
That's a good point. I think we had used this strategy for the FFT tests too, which is why I applied it here. Have pushed a change to not use skipping. Please let me know what you think.